### PR TITLE
Introduce ConceptCache

### PR DESCRIPTION
**Pull request by JeroenDeDauw**

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
777, do I win something?

**Comment by JeroenDeDauw** [comment]
Just split some code so far. Found the deletion stuff was duplicated and got rid of that. Also returned some array() and instead of empty returns, which where wrong. Should not have changed the actual behavior though. Obvious follow up work still needed to actually clean up the code, though this can be merged if it passes as far as I'm concerned.

File: src/SQLStore/QueryEngine/ConceptCache.php [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+		$cid = $this->store->smwIds->getSMWPageID(
+			$concept->getDBkey(),
+			SMW_NS_CONCEPT,
+			'',
+			'',
+			false
+		);
+
+		$db->deleteConceptById( $cid );
```
Comment (mwjames) [inline src/SQLStore/QueryEngine/ConceptCache.php]: `$db->deleteConceptById( $cid );` -> `$this->deleteConceptById( $cid );`

File: includes/storage/SQLStore/SMW_SQLStore3_Queries.php [hunk 2]
```diff
@@ -93,113 +86,39 @@ public function __construct( SMWSQLStore3 $parentstore, $dbslave ) {
...
 
-			if ( $qid < 0 ) {
-				return;
-			}
+	/**
+	 * @param string $desctxt
+	 *
+	 * @return SqlQueryPart
+	 */
+	public function doSomeStuffAndReturnSomething( $desctxt ) {
```
Comment (mwjames) [inline includes/storage/SQLStore/SMW_SQLStore3_Queries.php]: doSomeStuffAndReturnSomething ??

File: src/SQLStore/QueryEngine/ConceptCache.php (hunk 1 above)
Comment (JeroenDeDauw) [inline src/SQLStore/QueryEngine/ConceptCache.php]: Yeah I just noticed, kinda fail :)

File: includes/storage/SQLStore/SMW_SQLStore3_Queries.php (hunk 2 above)
Comment (JeroenDeDauw) [inline includes/storage/SQLStore/SMW_SQLStore3_Queries.php]: It is used. I factored this code out from ConceptCache::refresh.

You're right the name is unclear. For me to give it a more meaningful name I first need to understand what the code actually does.

Comment (mwjames) [inline includes/storage/SQLStore/SMW_SQLStore3_Queries.php]: `$desctxt` contains the raw `[[Modification date::+]]` query and the code within `doSomeStuffAndReturnSomething` is to resolve the query tree and return a sql transformation which is to be used to build the final joins/from/where sql condition.

Comment (mwjames) [inline includes/storage/SQLStore/SMW_SQLStore3_Queries.php]: Maybe `resolveQueryTreeForQueryCondition` (QueryCondition being the textual representation of the concept description)

Comment (JeroenDeDauw) [inline includes/storage/SQLStore/SMW_SQLStore3_Queries.php]: Renamed

**Comment by JeroenDeDauw** [comment]
Tests ought to pass now

**Comment by mwjames** [comment]
:+1:

**Comment by mwjames** [comment]
Follow-up #779

**Pull request merged**